src/sha256.nr

```diff
@@ -512,6 +485,173 @@ fn hash_final_block(msg_block: MSG_BLOCK, mut state: STATE) -> HASH {
     out_h
 }
 
+fn finalize_sha256_blocks<let N: u32>(
+    msg: [u8; N],
+    message_size: u32,
+    total_len: u32,
+    mut h: STATE,
+    mut msg_block: MSG_BLOCK,
+    mut msg_byte_ptr: u32,
+) -> HASH {
+    let modulo = total_len % BLOCK_SIZE;
+    // Handle setup of the final msg block.
+    // This case is only hit if the msg is less than the block size,
+    // or our message cannot be evenly split into blocks.
+    if modulo != 0 {
+        let num_blocks = total_len / BLOCK_SIZE;
+        let msg_start = BLOCK_SIZE * num_blocks;
+        let (new_msg_block, new_msg_byte_ptr) =
+            unsafe { build_msg_block(msg, message_size, msg_start) };
+
+        if msg_start < message_size {
+            msg_block = new_msg_block;
+        }
+
+        let new_msg_byte_ptr = verify_msg_block(msg, message_size, msg_block, msg_start);
+        if msg_start < message_size {
+            msg_byte_ptr = new_msg_byte_ptr;
+            verify_msg_block_padding(msg_block, msg_byte_ptr);
+        }
+    }
+
+    // If we had modulo == 0 then it means the last block was full,
+    // and we can reset the pointer to zero to overwrite it.
+    if msg_byte_ptr == BLOCK_SIZE {
+        msg_byte_ptr = 0;
+    }
+
+    // Pad the rest such that we have a [u32; 2] block at the end representing the length
+    // of the message, and a block of 1 0 ... 0 following the message (i.e. [1 << 7, 0, ..., 0]).
+    // Here we rely on the fact that everything beyond the available input is set to 0.
+    let index = msg_byte_ptr / INT_SIZE;
+    msg_block[index] = set_item_byte_then_zeros(msg_block[index], msg_byte_ptr, 1 << 7);
+
+    msg_byte_ptr = msg_byte_ptr + 1;
+    let last_block = msg_block;
+
+    // If we don't have room to write the size, compress the block and reset it.
+    if msg_byte_ptr > MSG_SIZE_PTR {
+        h = sha256_compression(msg_block, h);
+
+        // `attach_len_to_msg_block` will zero out everything after the `msg_byte_ptr`.
+        msg_byte_ptr = 0;
+    }
+
+    msg_block = unsafe { attach_len_to_msg_block(msg_block, msg_byte_ptr, message_size) };
+
+    verify_msg_len(msg_block, last_block, msg_byte_ptr, message_size);
+
+    hash_final_block(msg_block, h)
+}
+
+/**
+ * Partially computes a SHA256 hash of a message but does not finalize
+ * @notice can be used for post-partial hashing where client proves part of hash and relies on server to finish
+ * 
+ *
+ * @param N: the length of the message to hash.
+ *  --- WARNING: N must be divisible by BLOCK_SIZE such that N % BLOCK_SIZE == 0 
+ *      otherwise the remaining bytes will not be inputted when computing the initial hash
+ * @param msg: the preimage to begin hashing
+ * @return the intermediate hash state
+ */
+pub fn partial_sha256_var_start<let N: u32>(msg: [u8; N], message_size: u32) -> [u32; 8] {
```

Comment:
Looks like this should be just making a call to `partial_sha256_var_interstitial` while passing `INITIAL_STATE`